rust/lance/src/index/vector/ivf/builder.rs

```diff
@@ -0,0 +1,218 @@
+    data: impl RecordBatchStream + Unpin,
+    column: &str,
+    ivf: &mut Ivf,
+    pq: &ProductQuantizer,
+    metric_type: MetricType,
+    part_range: Range<u32>,
+) -> Result<()> {
+    let schema = data.schema();
+    if schema.column_with_name(column).is_none() {
+        return Err(Error::Schema {
+            message: format!("column {} does not exist in data stream", column),
+            location: location!(),
+        });
+    }
+    if schema.column_with_name(ROW_ID).is_none() {
+        return Err(Error::Schema {
+            message: "ROW ID is not set when building index partitions".to_string(),
+            location: location!(),
+        });
+    }
+
+    let ivf_immutable = Arc::new(ivf.clone());
+    let mut stream = data
+        .zip(repeat_with(|| (part_range.clone(), ivf_immutable.clone())))
+        .map(move |(b, (range, ivf_ref))| async move {
+            let batch = b?;
+            let col = column.to_string();
+            let range_copy = range.clone();
+            // Filter out the rows that are not in the partition range.
+            let batch = tokio::task::spawn_blocking(move || {
+                filter_batch_by_partition(&batch, &col, &ivf_ref, metric_type, range_copy)
+            })
+            .await??;
+            // Run product quantization.
+            let arr = batch
+                .column_by_name(RESIDUAL_COLUMN)
+                .expect("The caller already checked column exist")
+                .as_fixed_size_list();
+            let data = MatrixView::<Float32Type>::new(
+                Arc::new(arr.values().as_primitive::<Float32Type>().clone()),
+                arr.value_length() as usize,
+            );
+            let pq_code = pq.transform(&data, metric_type).await?;
+            let pq_field = Field::new(PQ_CODE_COLUMN, pq_code.data_type().clone(), false);
+            let batch = batch.try_with_column(pq_field, Arc::new(pq_code))?;
+            // Do not need to serialize original vector
+            let batch = batch.drop_column(column)?.drop_column(RESIDUAL_COLUMN)?;
+            Ok::<(arrow_array::RecordBatch, std::ops::Range<u32>), Error>((batch, range))
+        })
+        .buffer_unordered(num_cpus::get())
+        .and_then(|batch_and_range| async move {
+            // Split batch into per-partition batches
+            let (batch, range) = batch_and_range;
+            Ok(stream::iter(range).map(move |part_id| {
+                let predictions = BooleanArray::from_unary(
+                    batch
+                        .column_by_name(PARTITION_ID_COLUMN)
+                        .unwrap()
+                        .as_primitive::<UInt32Type>(),
+                    |pid| pid == part_id,
+                );
+                let parted_batch =
+                    filter_record_batch(&batch, &predictions)?.drop_column(PARTITION_ID_COLUMN)?;
+                Ok::<(u32, RecordBatch), Error>((part_id, parted_batch))
+            }))
+        })
```

Comment:
This is CPU bound, so I think we need to put this into `tokio::task::spawn_blocking()` to get any concurrency.

Reply:
I changed this piece of code.